### DISCUSSION TITLE
Inplace matrix operations

### DISCUSSION THREAD

**Discussion by JoaquinKeller** [josdejong/mathjs#2669] [Design decisions]
Hello ! 
I am a newb in mathjs and this is my first time posting here... 

My question:
When dealing with big matrices allocating memory can be costly. 
Using subset I wrote a piece of code (see below) that allows to assign a new value in an already allocated matrix. 
I was wondering is there is another way of doing so. 

Also it would be great if operations could be done inplace. example
```javascript
let m = math.ones(10**8);
m.add(4).multiply(2); // m would be now full of 10 (this feature does not exists)
```
Or maybe there is a way of doing inline operations and I haven't figure it out....
 
```javascript  
/**
 * [put a value inside a mathjs matrix without reallocating memory]
 * @param  {[math.Matrix]} m [a matrix]
 * @param  {[math.Matrix]} value [a matrix with same shape as m]
 * @return {[math.Matrix]}     [the matrix m with a new value]
 */
function putinside(m, value) {
    const s = m.size();
    let index = s.map(i => math.range(0, i));
    m.subset(math.index(...index), value);
    return m; // m hasn't changed, so this is not really needed
}
```

**Comment by josdejong** [comment]
Thanks for your question Joaquin, it's a good question. mathjs is designed with immutability in mind, so all operations like `add` and `multiply` do not alter the inputs but return a new matrix instance containing the results. Working in an immutable is a good idea in general since it can prevent quite some bugs and weird side effects. 

For example [moment.js](https://momentjs.com/) has this behavior where doing `m.add(...)` changes `m` itself. This has cost me quite some headaches in the past where this leads to unwanted side effects like:

```js
const start = moment()
const end = start.add(1, 'day') // whoops I accidentally changed start too!
```
So in _general_ I prefer immutable solutions.

Using methods on a matrix instead which mutate the matrix itself can indeed improve performance when working with large matrices, so there it can be a very valid case. I doubt though whether it would be a good idea to combine both models (mutable and immutable) in the same library, that can lead to confusion I'm afraid. 

If you have a limited set of operations you want to perform on the large matrices you have, you could maybe extend the `DenseMatrix` class with these methods yourself, like create a method `add` (or `addInplace`) on the prototype.

**Comment by JoaquinKeller** [comment]
Thanks for the answer. 
Actually mathjs is quite complete. I've been comparing with alternatives for linear algebra in js (ndarray and tensorflowjs) and mathjs is the only one to implement complex numbers (tfjs support for complex numbers is poor and buggy, ndarray has none). 
So I've been able to achieve what I was looking for: a quantum computer simulateur to run in the browser --like in here: [Quantum Perceptron](https://qml.entropicalabs.io/)
However I am concerned by the speed. 
Have you thought of BLAS and WebAssembly  ? A 100x speed up should be possible by using the SIMD parallelism or gpu (when available)...

And yes, you are right: the inplace operations should be explicit, to avoid confusion.

**Comment by josdejong** [comment]
Great to hear mathjs is such a good fit for your application.

One area where there is much room for improvement in mathjs is performance (even without considering mutating matrices themselves). If you like I can email you a paper Eric and I wrote about mathjs a couple of years ago, which also has a section about performance. There are also some issues here in the issues section discussion options to improve performance.
